src/state-model.test.js

```diff
@@ -0,0 +1,164 @@
+                eventWatcher2 = sinon.fake();
+            });
+
+            afterEach(function() {
+                instance.off().stopListening();
+            });
+
+            it('triggers set: and enter: events for new attributes', function() {
+                instance.on('set:document', eventWatcher1);
+                instance.on('enter:document', eventWatcher2);
+                instance.set('document', 1);
+                assert(eventWatcher1.calledWith(instance, 1));
+                assert(eventWatcher2.calledWith(instance, 1));
+                assert(eventWatcher1.calledBefore(eventWatcher2));
+            });
+
+            it('triggers exit: and enter: events for changed values', function() {
+                instance.set('document', 1);
+                instance.on({
+                    'exit:document': eventWatcher1,
+                    'enter:document': eventWatcher2
+                });
+                instance.set('document', 2);
+                assert(eventWatcher1.calledWith(instance, 1));
+                assert(eventWatcher2.calledWith(instance, 2));
+                assert(eventWatcher1.calledBefore(eventWatcher2));
+            });
+
+            it('triggers exit: and unset: for removed attributes', function() {
+                instance.set('document', 2);
+                instance.on({
+                    'exit:document': eventWatcher1,
+                    'unset:document': eventWatcher2
+                });
+                instance.unset('document');
+                assert(eventWatcher1.calledWith(instance, 2));
+                assert(eventWatcher2.calledWith(instance, 2));
+                assert(eventWatcher1.calledBefore(eventWatcher2));
+            });
+
+            it('can handle complex mixtures of changes', function() {
+                // We are going to use separate event listeners for all possible
+                // combinations of event type and attribute name. There are four
+                // event types and in this test, we track three attributes, so
+                // twelve listeners in total.
+                var watchers = [eventWatcher1, eventWatcher2]
+                    .concat(_.times(10, function() { return sinon.fake(); }));
+                var attributeNames = ['document', 'page', 'user'];
+                var eventNames = ['set:', 'exit:', 'enter:', 'unset:'];
+                // Attributes before we start listening for events.
+                instance.set({document: 1, page: 10});
+                // Backbone.Model supports recursive changes during model change
+                // events, and so do we. We mimic this below by unsetting the
+                // `page` attribute while the `document` attribute is being
+                // updated. This also enables us to capture simultaneous `set:`
+                // and `unset:` events in a single "burst" of changes.
+                instance.once('change:document', instance.unset.bind(instance, 'page'));
```

Comment:
I haven't seen `Function.bind()` in years, maybe an arrow function would be easier for the reader?

Reply:
I feel that `.bind` is more correct in this case, because it guarantees that the new function behaves exactly like the old function. Besides, recognizability is subjective.

(Also, if this were not a test, I would avoid arrow functions anyway because I am trying not to force myself or the users to add a transpilation step.)